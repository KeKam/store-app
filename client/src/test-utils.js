import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
