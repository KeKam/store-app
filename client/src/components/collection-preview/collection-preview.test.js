import React from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router, Route } from 'react-router-dom';

import { render, screen } from '../../test-utils';
import CollectionPreview from './collection-preview';

const history = createMemoryHistory({ initialEntries: ['/collection'] });

describe('CollectionPreview component', () => {
  const mockCollectionProps = {
    id: 1,
    title: 'SPRING',
    items: [
      {
        id: 1,
        imageUrl: 'www.testImage.com',
        name: 'Tulips',
        price: 20,
      },
    ],
    routeName: 'spring',
  };

  it('should render CollectionPreview component', () => {
    const { asFragment } = render(
      <Router history={history}>
        <CollectionPreview {...mockCollectionProps} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole('heading', { name: 'SPRING' })).toBeInTheDocument();
    expect(screen.getByTestId('image')).toBeInTheDocument();
    expect(screen.getByTestId('title')).toHaveTextContent('Tulips');
    expect(screen.getByTestId('price')).toHaveTextContent('20 â‚¬');
    expect(screen.getByTestId('link')).toHaveTextContent('Sign in required');
  });

  it('should history push on click', () => {
    render(
      <Router history={history}>
        <Route path='/collection'>
          <CollectionPreview {...mockCollectionProps} />
        </Route>
      </Router>
    );

    history.push = jest.fn();

    userEvent.click(screen.getByRole('heading', { name: 'SPRING' }));

    expect(history.push).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledWith('/collection/spring');
  });
});
