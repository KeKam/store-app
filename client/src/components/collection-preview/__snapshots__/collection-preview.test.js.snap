// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionPreview component should render CollectionPreview component 1`] = `
<DocumentFragment>
  <div
    class="sc-dnqmqq dZJyys"
  >
    <h1
      class="sc-iwsKbI cHIACp"
    >
      SPRING
    </h1>
    <div
      class="sc-gZMcBi lkZMND"
    >
      <div
        class="sc-bwzfXH srITS"
      >
        <div
          class="sc-htpNat dTlvyq"
          data-testid="image"
        />
        <div
          class="sc-bxivhb oDKzY"
        >
          <span
            class="sc-EHOje ckFQmK"
            data-testid="title"
          >
            Tulips
          </span>
          <span
            class="sc-bZQynM bjXVPL"
            data-testid="price"
          >
            20 â‚¬
          </span>
        </div>
        <a
          class="sc-gzVnrw jnQMmI"
          data-testid="link"
          href="/login"
        >
          <h2
            class="sc-htoDjs bRjjzo"
          >
            Sign in required
          </h2>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
